### PR TITLE
Add Initial Implementation of Correctly Sized Header Vectors for Restart

### PR DESCRIPTION
**Pull request by atgeirr**
This is another PR for the same thing, it is #392 with an extra commit documenting some literals as previously discussed. It is marked WIP, as I get local test failures. I'll run jenkins anyway to see if it's only my system, but the PR should not be merged until I have resolved those failures.

### CODE REVIEW

**Comment by atgeirr** [comment]
jenkins build this please

File: src/opm/output/eclipse/InteHEAD.cpp [hunk 1]
```diff
@@ -633,4 +642,66 @@ Opm::RestartIO::InteHEAD::regionDimensions(const RegDims& rdim)
...
+        return system_clock::to_time_t(t);
+    }
+}
+
+std::time_t
+Opm::RestartIO::makeUTCTime(const std::tm& timePoint)
+{
+    auto       tp    =  timePoint; // Mutable copy.
+    const auto ltime =  std::mktime(&tp);
+    auto       tmval = *std::gmtime(&ltime); // Mutable.
```
Comment (atgeirr) [inline src/opm/output/eclipse/InteHEAD.cpp]: The call to gmtime() is not thread-safe, which can *in theory* cause problems as we actually use multiple threads for output. However, as the output writes are pipelined it is probably ok (if not, the restart file would be corrupted, which would be much more noticable than a wrong data somewhere...).

Comment (bska) [reply src/opm/output/eclipse/InteHEAD.cpp]: >The call to `gmtime()` is not thread-safe,

Correct, and that's why we dereference the return value as soon as possible to get it into a local variable rather than relying on the internal static storage.  There will indeed be timing issues if multiple threads (from the same process) run this code.  If that's possible then we need to add some kind of synchronisation here.

Comment (bska) [reply src/opm/output/eclipse/InteHEAD.cpp]: Ideally we would have the `date` library from C++20....

**Comment by bska** [comment]
As for the test failure, commit ef29518 switches `InteHEAD::data_[ih_076]` from `2` to `5` in `InteHEAD::variousParam()` without updating the associate unit test.

File: src/opm/output/eclipse/DoubHEAD.cpp [hunk 2]
```diff
@@ -0,0 +1,603 @@
...
+        tm1.tm_year = 1;        // 1901
+        tm1.tm_mon  = tm0.tm_mon;
+        tm1.tm_mday = tm0.tm_mday;
+
+        tm1.tm_hour  = 1;
+        tm1.tm_min   = 0;
+        tm1.tm_sec   = 0;
+        tm1.tm_isdst = 0;
+
+        const auto t1 = ::Opm::RestartIO::makeUTCTime(tm1);
```
Comment (atgeirr) [inline src/opm/output/eclipse/DoubHEAD.cpp]: After quite a bit of reading docs and debugging, I have at last found a bug here: makeUTCTime() cannot be called with time points before the epoch, since then the mktime() call inside will return -1. That is not accounted for. The test failure I get is:
```
108: /Users/atgeirr/opm/opm-common/tests/test_DoubHEAD.cpp:88: error: in "Member_Functions/Time_Stamp": difference{0.000133156} between v[161 - 1]{735979} and 736077.0{736077} exceeds 1e-10%
108: /Users/atgeirr/opm/opm-common/tests/test_DoubHEAD.cpp:91: error: in "Member_Functions/Time_Stamp": difference{0.000133134} between v[162 - 1]{736102} and 736200.0{736200} exceeds 1e-10%
108: 
108: *** 2 failures are detected in the test module "DoubHEAD_Vector"
```
(note that line numbers have changed a little from this version).

Comment (atgeirr) [reply src/opm/output/eclipse/DoubHEAD.cpp]: Another issue: I am not at all sure if the ECL concept accounts for time zones or not. When the deck says:
```
START
9 APR 2015 /
```
I think we should assume no DST for example. The time point of the test (which not coincidentally is the date above) is given as a std::tm with tm_isdst = 0, but on that date in my timezone DST was active, so I get a UTC time point of April 8, 2300, which probably is why I get 98 days off instead of 99 days (which is what I would expect from the bug described above).

TL;DR: we should probably be totally timezone-agnostic in all C library usage. We should assume that we operate on UNIX epoch (which is guaranteed for system_clock from C++20 but in practice I think it is true for all systems), and never involve dates before that. The offset from year 0 to 1970 (in "ECL calendar") is a fixed number of days that we could store.

**Comment by atgeirr** [comment]
> As for the test failure, commit ef29518 switches InteHEAD::data_[ih_076] from 2 to 5 in InteHEAD::variousParam() without updating the associate unit test.

I have fixed that locally. If only the locally failing test was as easy...

**Comment by atgeirr** [comment]
jenkins build this

**Comment by atgeirr** [comment]
jenkins build this

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
The makeUTCTime() function is essentially an attempt at a portable timegm() (see https://www.gnu.org/software/libc/manual/html_node/Broken_002ddown-Time.html). I think that it is good as long as we stay ahead of epoch.

**Comment by atgeirr** [comment]
jenkins build this please

**Comment by atgeirr** [comment]
@bska Please merge when green, if satisfied with the solution.
